nilearn/_utils/param_validation.py

```diff
@@ -361,3 +363,85 @@ def _cast_to_int32(sample_mask):
             what can be represented by int32: {highest}."
         raise ValueError(msg)
     return np.asarray(sample_mask, new_dtype)
+
+
+def check_params(fn_dict):
+    """Check types of inputs passed to a function / method / class.
+
+    This function checks the types of function / method parameters or
+    the attributes of the class.
+
+    This function is made to check the types of the parameters
+    described in ``nilearn._utils.docs``
+    that are shared by many functions / methods / class
+    and thus ensure a generic way to do input validation
+    in several important points in the code base.
+
+    In most cases this means that this function can be used
+    on functions / classes that have the ``@fill_doc`` decorator,
+    or whose doc string uses parameter templates
+    (for example ``%(data_dir)s``).
```

Comment:
So IIUC, for this check to work, would it be enough to write the expected type of parameters in the docstring? 

If so, is there some syntax that needs to be followed or simply using the `:obj:` rst tags is enough.

(I might be getting this completely wrong, so please correct me if that's the case)

Reply:
> would it be enough to write the expected type of parameters in the docstring?

not quite: I wished it were that powerful

to expand the functionality of the function you need to:
- describe the expected type for that parameter / attribute in `nilearn.typing` (it must be something that `isinstance` can handle
- expand the `type_map` dictionary of `check_params` top pair the name of the parameter / attribute with its expected type

does that help?

Reply:
I see. So, everywhere we call this, we only check the type of a _subset_ of parameters defined in the function.

And, this subset consists of the parameters we have in `nilearn._utils.docs`/

Reply:
this subset consists of **SOME** of the parameters we have in `nilearn._utils.docs`

ideally I would like to cover all of them but at the moment it is not the case

Reply:
> to expand the functionality of the function you need to:

I was just thinking that it might be good to document this in some way? Maybe the contributing page? But since it isn't full-fledged typing maybe even top of the `typing.py` would suffice?

Reply:
will do in a follow up PR to keep this one under 500 lines changed